src/nvim/lua/treesitter.c

```diff
@@ -1043,13 +1046,17 @@ static int query_next_capture(lua_State *L)
     lua_pushinteger(L, capture.index+1);  // [index]
     push_node(L, capture.node, lua_upvalueindex(2));  // [index, node]
 
+    // Now check if we need to run the predicates
     uint32_t n_pred;
     ts_query_predicates_for_pattern(query, match.pattern_index, &n_pred);
-    if (n_pred > 0 && capture_index == 0) {
+
+    if (n_pred > 0
+        && (capture_index == 0 || ud->current_pattern != match.pattern_index)) {
```

Comment:
Hmm after some thinking this not either correct. only `match.id` is unique per match, and matches could overlap arbitrarily. Ideally, tree-sitter should tell us when it is the first time a certain capture cursor is seeing a certain match. I need to check what the official ts highlight module is doing.

Reply:
after reading a bit of tree-sitter query.c internals, I think the easiest solution is a patch to libtreesitter which simply adds a bool flag to the TSQueryMatch objects which informs the caller whether this is the first time that match is being returned.

Reply:
I am not sure this is exactly easy.
The fact that matches can overlap is a pain already, maybe we should just keep track of the `match.id` we saw during the execution of the query ?

In libtreesitter, from what I see the only valuabl piece of information we could get is from `QueryState.consumed_capture_count` which itself is not great because it may be `>0` even when the match is first returned, when there is captures that have been skipped because they are before the cursor's start byte.

Reply:
Ans there's another issue, that points us again in the direction of either using tables to store results everywhere, or sometimes rerunning the predicates: when matches overlap, the library will return the first match, which we will use to set the lua table that we return, then fill with whether this match is active or not. On the next match, we will do the same, and overwrite the `active` value that we computed. Then if the first match comes back (because of overlap), we will have to recompute wether this match is active or not.

Reply:
I was thinking of just adding a `bool : 1` to state which indicates whether it has been returned as a `TSQueryMatch` yet.

Reply:
Yes but even if we do that, we might have to recompute things anyway, because of overlapping matches.

Reply:
Hmm i don't think so. The internal QueryState object should map 1:1 to a match, right? I can open a PR to tree sitter and see what Max thinks about it.

Reply:
I see what you mean, if we encounter a capture from a match already encountered, we know that the predicate evaluation was `true`, because it would have been removed otherwise.